### PR TITLE
Incorporated SVD

### PR DESCRIPTION
**Pull request by AndPotap**
Incorporated SVD functionality into CoLA.

### CODE REVIEW

File: cola/linalg/decompositions/decompositions.py [hunk 1]
```diff
@@ -1,20 +1,40 @@
...
+    ScalarMul,
+    Triangular,
+)
+from cola.utils import export
 
 PRNGKey = Any
 
 
+@export
+class SVD(Algorithm):
```
Comment (mfinzi) [inline cola/linalg/decompositions/decompositions.py]: SVD isn't really an algorithm though right? Is this meant to refer to the dense method for computing it? At the very least we should probably have a different name

File: cola/linalg/decompositions/decompositions.py [hunk 2]
```diff
@@ -78,6 +98,31 @@ def __call__(self, A: LinearOperator):
         return arnoldi(A, **self.__dict__)
 
 
+@dataclass
```
Comment (mfinzi) [inline cola/linalg/decompositions/decompositions.py:91]: why not just use the Lanczos algorithm? How lanczos gets used should be in the dispatch for SVD based on the alg I think?

File: cola/linalg/svd/svd.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    """
+
+
+@dispatch(precedence=-1)
+def svd(A: LinearOperator, alg: Auto):
+    """ Auto:
+        - if A is small, use dense SVD
+        - if A is large, use dense Lanczos SVD
+    """
+    if not hasattr(alg, "method"):
```
Comment (mfinzi) [inline cola/linalg/svd/svd.py]: I would avoid logic looking like this (the kind of string based handling we had before). Instead switch based on the types, see the auto for inv or logdet or eigs

File: cola/linalg/decompositions/decompositions.py [hunk 4]
```diff
@@ -78,6 +98,31 @@ def __call__(self, A: LinearOperator):
...
+            PRNGKey for jax, long integer for numpy or pytorch.
+    """
+    start_vector: Array = None
+    max_iters: int = 1_000
+    tol: float = 1e-6
+    pbar: bool = False
+    key: Optional[PRNGKey] = None
+
+    def __call__(self, A: LinearOperator):
+        Sigma0_2, V, _ = lanczos_eigs(A.H @ A, **self.__dict__)
```
Comment (mfinzi) [inline cola/linalg/decompositions/decompositions.py:110]: I think you can do it with one call, get V and then A@V should give the other singular vectors (when normalized)

**Review by mfinzi — CHANGES_REQUESTED** [review]
I would probably remove the algorithms entirely (and instead just reuse the 'lanczos' algorithm like we would e.g. in logdet). Then for the auto method, I would make it more paradigmatic. Have the real work (for converting lanczos outputs to U S V done in the dispatched svd(A, alg) call, most similar to how we do it with the other functions.

**Comment by mfinzi** [comment]
We should also think carefully about what the arguments of SVD should be, just like for Eigs. We definitely can't get away with just SVD(A,alg)

File: cola/linalg/decompositions/decompositions.py (hunk 1 above)
Comment (AndPotap) [reply cola/linalg/decompositions/decompositions.py]: Wouldn't that apply for the LU algorithm? I based the choice of having creating `SVD` mimicking what we have for `Cholesky` and `LU.` Should we also change say `LU` to Gaussian Elimination?

File: cola/linalg/decompositions/decompositions.py (hunk 2 above)
Comment (AndPotap) [reply cola/linalg/decompositions/decompositions.py:91]: Could you next time encapsulate the code chuck and not just a single line. I wasn't sure what you were referring to.
To your point, it makes sense. I'll do the change.

File: cola/linalg/svd/svd.py (hunk 3 above)
Comment (AndPotap) [reply cola/linalg/svd/svd.py]: I'm unsure that we want types here. If we know the operator is PSD, why would we do SVD? Or do you think that the svd function should dispatch to the Lanczos case for those types?

File: cola/linalg/decompositions/decompositions.py (hunk 4 above)
Comment (AndPotap) [reply cola/linalg/decompositions/decompositions.py:110]: Good point.

**Comment by AndPotap** [comment]
> We should also think carefully about what the arguments of SVD should be, just like for Eigs. We definitely can't get away with just SVD(A,alg)

We are getting away already. I think you are hinting at the fact that we might want to improve the underlying algorithms and those might require more arguments. I'll add the arguments given that, right now I wouldn't know what to add.

File: cola/linalg/decompositions/decompositions.py (hunk 1 above)
Comment (AndPotap) [reply cola/linalg/decompositions/decompositions.py]: Also I don't think `Eig()` or `Eigh()` is an algorithm either.

**Review by mfinzi — APPROVED** [review]

**Pull request merged**